pkg/action/archive.go

```diff
@@ -0,0 +1,206 @@
+package action
+
+import (
+	"archive/tar"
+	"archive/zip"
+	"compress/gzip"
+	"errors"
+	"fmt"
+	"io"
+	"log"
+	"os"
+	"path"
+	"path/filepath"
+	"strings"
+
+	"github.com/ulikunitz/xz"
+)
+
+const maxBytes = 1 << 29 // 512MB
+
+func copyArchive(src string, dst string) error {
+	r, err := os.Open(src)
+	if err != nil {
+		return fmt.Errorf("failed to open source file: %w", err)
+	}
+	defer r.Close()
+
+	w, err := os.CreateTemp(dst, fmt.Sprintf("%s.*", filepath.Base(src)))
+	if err != nil {
+		return fmt.Errorf("failed to create temporary file: %w", err)
+	}
+
+	defer func() {
+		if cerr := w.Close(); cerr != nil {
+			log.Printf("failed to close file: %v", cerr)
+		}
+	}()
+
+	if _, err = io.Copy(w, r); err != nil {
+		return fmt.Errorf("failed to copy data: %w", err)
+	}
+
+	return nil
+}
+
+func tempDir(p string) (string, error) {
+	tmpDir, err := os.MkdirTemp("", fmt.Sprintf("bincapz-%s", filepath.Base(p)))
+	if err != nil {
+		return "", fmt.Errorf("failed to create temp dir: %w", err)
+	}
+
+	if err := copyArchive(p, tmpDir); err != nil {
+		os.RemoveAll(tmpDir)
+		return "", fmt.Errorf("failed to copy archive: %w", err)
+	}
+
+	return tmpDir, nil
+}
+
+func extractTar(d string, f string) error {
```

Comment:
This is essentially the `extract` function from #124 but modified to allow the other archive types.